libs2eplugins/src/s2e/Plugins/Models/BaseFunctionModels.cpp

```diff
@@ -692,42 +692,67 @@ bool BaseFunctionModels::strcatHelper(S2EExecutionState *state, const uint64_t s
-    getWarningsStream(state) << "Enter strstr yes\n";
+    getWarningsStream(state) << "Entering strstrHelper\n";
+
     size_t haystackLen, needleLen;
-    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) || !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
+    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) ||
+        !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
         getWarningsStream(state) << "Failed to find nullptr in haystack or needle\n";
         return false;
     }
 
-    const Expr::Width ret_width = state->getPointerSize() * CHAR_BIT;
+    const Expr::Width pointerWidth = state->getPointerSize() * CHAR_BIT;
+
     if (needleLen == 0) {
-        retExpr = E_CONST(haystackAddr, ret_width);
-        return true;
+        retExpr = E_CONST(haystackAddr, pointerWidth);
+        return true;  
     }
+    const ref<Expr> nullExpr = E_CONST(0, pointerWidth);
 
-    const ref<Expr> nullExpr = E_CONST(0, ret_width);
-
-    ref<Expr> finalExpr = nullExpr; // Start with the assumption that we will not find the needle
-    for (size_t i = 0; i <= haystackLen - needleLen; i += byte_width) {
-        uint64_t strAddrs[2] = {haystackAddr + i, needleAddr};
-        ref<Expr> memcmpResult;
-
-        // Use strncmpHelper to compare the substring of haystack with needle
-        if (!memcmpHelper(state, strAddrs, needleLen, memcmpResult)) {
-            getWarningsStream(state) << "Failed to compare substring of haystack with needle " << needleLen << "\n";
+    std::vector<size_t> badCharSkip(256, needleLen);  // Bad character skip array. initialized to needleLen for all entry
+    uint8_t charByte = 0;  // Temporary variable for reading bytes
+    for (size_t i = 0; i < needleLen - 1; ++i) {
+        uint64_t readAddr = needleAddr + i * byte_width;
+        if (!state->mem()->read(readAddr, &charByte, VirtualAddress, true)) {
+            getWarningsStream(state) << "Failed to read byte at address " << hexval(readAddr) << "\n";
             return false;
         }
+        badCharSkip[charByte] = needleLen - 1 - i; // set each of the needle char to corresponding skip len
+    }
+
+    size_t i = 0;  // Position in haystack
+    auto solver = state->solver();
+    while (i <= haystackLen - needleLen) {
+        uint64_t currentHaystackAddr = haystackAddr + i * byte_width;
+        uint64_t strAddrs[2] = {currentHaystackAddr, needleAddr};
+        ref<Expr> cmpResult;
+
+        if (strncmpHelper(state, strAddrs, needleLen, cmpResult)) {
+            // incrementally adding the state constriant based on solver feedback
+            ref<Expr> findNeedle = E_EQ(cmpResult, E_CONST(0, Expr::Int32));
+            Query query(state->constraints(), findNeedle);
+            bool truth;
+            if (solver->mayBeTrue(query,truth)) {
```

Comment:
Why this is a `maybeTrue`? So we stop at the potential match, right?

Reply:
yes. mustBeTrue may be too strict